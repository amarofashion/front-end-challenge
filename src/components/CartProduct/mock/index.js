export default {
  name: 'VESTIDO TRANSPASSE BOW',
  style: '20002605',
  onSale: false,
  regularPrice: 'R$ 199,90',
  actualPrice: 'R$ 199,90',
  discountPercentage: '',
  installments: '3x R$ 66,63',
  image:
    'https://d3l7rqep7l31az.cloudfront.net/images/products/20002605_615_catalog_1.jpg?1460136912',
  sizes: [
    {
      available: false,
      size: 'PP',
      sku: '5807_343_0_PP',
    },
    {
      available: true,
      size: 'P',
      sku: '5807_343_0_P',
    },
    {
      available: true,
      size: 'M',
      sku: '5807_343_0_M',
    },
    {
      available: true,
      size: 'G',
      sku: '5807_343_0_G',
    },
    {
      available: false,
      size: 'GG',
      sku: '5807_343_0_GG',
    },
  ],
  selectedSize: {
    available: false,
    size: 'PP',
    sku: '5807_343_0_PP',
  },
  quantity: 1,
  quantityCallback: item => console.log(item),
  deleteCallback: item => console.log(item),
};
