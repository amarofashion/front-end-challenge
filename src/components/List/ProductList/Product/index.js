import React, { Fragment } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Popup from "reactjs-popup";
import uuid from "uuid";

import Thumb from "../../../Thumb";

import { addProduct } from "../../../../api/redux/reducers/cart/actions";

const Product = ({ product, addProduct }) => {
  product.quantity = 1;

  const productInstallment = (
    <div className="installment">
      <span>ou {product.installments}</span>
    </div>
  );

  return (
    <div className="product-item" data-sku={product.sku}>
      {product.on_sale && (
        <div className="product-promotion">
          {product.discount_percentage && (
            <small>{product.discount_percentage} off</small>
          )}
        </div>
      )}
      <Thumb
        classes="product-item__thumb"
        src={product.image}
        alt={product.title}
      />
      <p className="product-item__title">{product.name}</p>

      <div className="product-item__price">
        <div className="val">
          {product.discount_percentage && (
            <span>De: {product.regular_price} Por:</span>
          )}
          <b>{product.actual_price}</b>
        </div>
        {productInstallment}
      </div>
      <Popup
        modal
        trigger={<button className="product-item__buy-btn">Adicionar</button>}
      >
        {close => (
          <div>
            Por favor selecione uma opção:
            <div className="product-item__size" onClick={close}>
              {product.sizes.map(value => {
                return (
                  <Fragment key={uuid.v4()}>
                    <button
                      disabled={!value.available}
                      onClick={() =>
                        addProduct({
                          ...product,
                          id: value.sku,
                          variant: value.size
                        })
                      }
                    >
                      {value.size}
                    </button>
                  </Fragment>
                );
              })}
            </div>
          </div>
        )}
      </Popup>
    </div>
  );
};

Product.propTypes = {
  product: PropTypes.object.isRequired,
  addProduct: PropTypes.func.isRequired
};

export default connect(
  null,
  { addProduct }
)(Product);
