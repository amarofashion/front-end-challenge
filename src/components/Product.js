import React from 'react'
import Size from './Size'
import { Col, Image, Button, Glyphicon } from 'react-bootstrap'

const { string, number, bool, arrayOf, shape, func } = React.PropTypes

class Product extends React.Component {

  constructor(props) {
    super(props)
    this.state = { size: this.props.sizes.filter(s => s.available)[0] }
  }

  render() {

    let {
      name,
      image,
      defaultImage,
      sizes,
      on_sale,
      actual_price,
      regular_price,
      onAddToCart } = this.props

    let { size } = this.state

    let handleAddToCartClick = () => {
      onAddToCart({
        id: size.sku,
        name,
        image,
        price: on_sale ? actual_price : regular_price,
        size
      })
    }

    let handleSizeClick = size => {
      this.setState({ size })
    }

    return (
      <Col md={3} className="product">
        <Image src={ image || defaultImage } responsive/>
        <h5 className="name">{ name.toLowerCase() }</h5>
        <div> { on_sale ?
          <span>
            de <span className="price-scratched">
              { regular_price }</span> por <span className="price">
              { actual_price }</span>
          </span> :
          <span className="price">
            { regular_price }
          </span> }
        </div>
        <div>
          <Size
            sizes={ sizes
              .filter( s => s.available )
              .map( s => ({ ...s, id: s.sku }) ) }
            onClick={ handleSizeClick } />
        </div>
        <Button
          bsStyle="primary"
          onClick={ handleAddToCartClick }>
            <Glyphicon glyph="shopping-cart" /> Adicionar ao carrinho
        </Button>
      </Col> ) }

}

Product.PropTypes = {
  name: string.isRequired,
  image: string.isRequired,
  defaultImage: string.isRequired,
  regular_price: number.isRequired,
  actual_price: number.isRequired,
  on_sale: bool.isRequired,
  sizes: arrayOf(
    shape({
      available: bool.isRequired,
      size: string.isRequired,
      sku: string.isRequired
    }).isRequired
  ),
  onAddToCart: func.isRequired
}

export default Product