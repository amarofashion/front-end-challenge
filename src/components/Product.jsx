import React from 'react';
import { Component } from 'react';
import PropTypes from 'prop-types';

export default class Product extends Component {
	static propTypes = {
		toCart: PropTypes.func
	}

	static defaultProps = {
		toCart: () => {}
	}

	constructor(props) {
		super(props);

		this.toCart = this.toCart.bind(this);
	}

	renderName({name}) {
		return <h2 className="name">{ name }</h2>;
	}

	renderPrice({actual_price, regular_price, discount_percentage}) {
		if(actual_price !== regular_price)
			return (<span className="price">
				De <del>{ regular_price }</del> por { actual_price } <em>({ discount_percentage } de desconto)</em>
			</span>);

		return <span className="price">{ actual_price || regular_price }</span>;
	}

	getAvailableSizes(sizes) {
		return sizes.filter(({available}) => available)
					.map(({size}) => size)
					.join(',');
	}

	renderSizes({sizes}) {
		if(!sizes || !sizes.length)
			return null;

		return <span className="sizes">Tamanhos dispon√≠veis: { this.getAvailableSizes(sizes) }</span>
	}

	renderImage({image, name}) {
		if(!image)
			return null;

		const alt = name ? `Foto de ${name}` : '';

		return <img src={ image } alt={ alt } className="image" />
	}

	toCart() {
		this.props.toCart(this);
	}

	renderCartButton() {
		return <button type="button" className="tocart" onClick={ this.toCart }>Adicionar ao carrinho</button>;
	}

	render() {
		const { props } = this;
		const onSale = props.on_sale ? 'onsale' : '';
		return (<div className={ `product-box ${onSale}` } id={ `p${props.id}` }>
			{ this.renderImage(props) }
			{ this.renderName(props) }
			{ this.renderPrice(props) }
			{ this.renderSizes(props) }
			{ this.renderCartButton() }
		</div>);
	}
}