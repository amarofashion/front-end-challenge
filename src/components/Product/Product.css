.product {
    margin: 0;
    overflow: auto;
}

.discount-percentage,
.regular-price,
.installments,
.product__price {
    display: inline-block;
    margin: 0 10px 10px 0;
    font-size: 12px;
}

.discount-percentage,
.regular-price {
    font-weight: bold;
    font-size: 15px;
}

.discount-percentage {
    color: #cf3838
}

.is-old {
    color: #999;
    text-decoration: line-through;
}

.product__button-wrap {
    display: block;
}

.product__button-wrap .btn-sale {
    background-color: #000;
    color: #FFF;
    font-size: 14px;
    text-transform: uppercase;
    width: 100%;
    text-align: center;
    padding: 10px;
    display: inline-block;
    border: none;
}

.product__button-wrap .btn-sale:disabled {
    background-color: #d0d0d0;
    cursor: not-allowed;
}

.box-prices {
    min-height: 82px;
}

.product-title {
    font-size: 14px;
    color: #000;        
    cursor: pointer;
    margin: 5px 0;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
}

.none {
    display: none;
}

.product-thumbnail {
    display: block;
    margin-bottom: 20px;
    line-height: 1.42857143;

    transition: border .2s ease-in-out;
}

.product-image {
    cursor: pointer;
}

.disable-image {
    opacity: 0.4;
}

.mdl-card__actions {
  display: flex;
  box-sizing:border-box;
  align-items: center;
}
.mdl-card__actions > .mdl-button--icon {
  margin-right: 3px;
  margin-left: 3px;
}

.mdl-card__media {
    background-color:transparent!important;
}

.button-buy {
    background: #000;
    color: #FFF;
    padding: 10px;
    display: inline-block;
    text-align: center;
}


.mdl-card__supporting-text {
    min-height: 70px;
}