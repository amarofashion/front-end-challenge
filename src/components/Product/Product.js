import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Product extends Component {

    handleClick = () => {
        const { id, addToCart, isInCart } = this.props;
        if(!isInCart) {
            addToCart(id);
        }
    }

    render() {
        const { name, actual_price, installments, regular_price, image, discount_percentage, isInCart } = this.props;

        return (
            <div className={isInCart ? "product-image disable-image" : "product-image"}>
                <figure className="mdl-card__media" onClick={this.handleClick}>
                    <img src={image} alt={name} title={name} />
                </figure>
                <div className="mdl-card__title">
                    <h1 className="product-title" onClick={this.handleClick}>{name}</h1>
                </div>
                <div className="mdl-card__supporting-text">
                    <p className={`product__price ${regular_price !== actual_price ? 'is-old' : ''}`}>{actual_price}</p>
                    <p className="product__price">{actual_price}</p> 
                    <p className="installments">{installments}</p>
                    {regular_price !== actual_price ?
                            <div>
                                <p className="regular-price">{regular_price}</p>
                                <p className="discount-percentage">({discount_percentage} off)</p>
                            </div>
                        : null
                    } 
                </div>
                <div className="mdl-card__actions mdl-card--border">
                    <a  onClick={this.handleClick} className="mdl-cell--12-col button-buy mdl-js-button">Comprar</a>
                </div>
            </div>
        );
    }
}

Product.propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    discount_percentage: PropTypes.string,
    regular_price: PropTypes.string,
    actual_price: PropTypes.string,
    installments: PropTypes.string,
    image: PropTypes.string,
    url: PropTypes.string,
    isInCart: PropTypes.bool.isRequired,
    addToCart: PropTypes.func.isRequired,
    removeFromCart: PropTypes.func.isRequired,
}

export default Product;

