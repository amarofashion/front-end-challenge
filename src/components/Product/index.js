import React from 'react';
import { string, bool, func, shape, arrayOf } from 'prop-types';

import { StyledWrapper } from './style';

const propTypes = {
  image: string.isRequired,
  name: string.isRequired,
  regularPrice: string.isRequired,
  actualPrice: string.isRequired,
  discountPercentage: string.isRequired,
  installments: string.isRequired,
  onSale: bool.isRequired,
  sizes: arrayOf(
    shape({
      available: bool,
      size: string,
      sku: string,
    })
  ).isRequired,
  callback: func.isRequired,
};

function Product({
  image,
  name,
  regularPrice,
  onSale,
  actualPrice,
  discountPercentage,
  installments,
  sizes,
  callback,
}) {
  return (
    <StyledWrapper onClick={() => callback()} sale={onSale}>
      <div className='content'>
        <div className='image'>
          {onSale && (
            <div className='sale'>
              <p className='title'>SALE</p>
              <p className='percentage'>{discountPercentage}</p>
            </div>
          )}

          <img src={image} alt={`Foto do produto ${name}`} />
        </div>

        <div className='bottomContent'>
          <p className='name'>{name}</p>

          {onSale
            ? [
                <p key='regularPrice' className='regularPrice prices'>
                  {regularPrice}
                </p>,
                <p key='actualPrice' className='actualPrice prices'>
                  {actualPrice}
                </p>,
                <p key='installments' className='installments prices'>
                  {installments}
                </p>,
              ]
            : [
                <p key='regularPrice' className='regularPrice prices'>
                  {regularPrice}
                </p>,
                <p key='installments' className='installments prices'>
                  {installments}
                </p>,
              ]}

          <div className='sizes'>
            {sizes.map(
              ({ available, size }, index) =>
                available && (
                  <div key={`size-${index}`} className='size'>
                    <span>{size}</span>
                  </div>
                )
            )}
          </div>
        </div>
      </div>
    </StyledWrapper>
  );
}

Product.propTypes = propTypes;

export default Product;
