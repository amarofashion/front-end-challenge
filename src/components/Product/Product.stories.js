import React from 'react';

import Product from '.';
import mock from './mock';

export default {
  component: Product,
  title: 'Product',
};

export const regular = () => (
  <div style={{ width: 400, margin: 10 }}>
    <Product {...mock} />
  </div>
);

export const onSale = () => (
  <div style={{ width: 400, margin: 10 }}>
    <Product {...mock} onSale />
  </div>
);
