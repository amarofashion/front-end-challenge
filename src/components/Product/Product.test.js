import React from 'react';
import { mount } from 'enzyme';

import Product from '.';
import mock from './mock';

describe('<Product />', () => {
  let wrapper;

  describe('with default props', () => {
    beforeAll(() => {
      wrapper = mount(<Product {...mock} />);
    });

    it('shoud match snapshot with default props', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
