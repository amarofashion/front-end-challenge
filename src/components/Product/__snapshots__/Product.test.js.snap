// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> with default props shoud match snapshot with default props 1`] = `
<Product
  actualPrice="R$ 199,90"
  callback={[Function]}
  discountPercentage=""
  image="https://d3l7rqep7l31az.cloudfront.net/images/products/20002605_615_catalog_1.jpg?1460136912"
  installments="3x R$ 66,63"
  name="VESTIDO TRANSPASSE BOW"
  onSale={false}
  regularPrice="R$ 199,90"
  sizes={
    Array [
      Object {
        "available": false,
        "size": "PP",
        "sku": "5807_343_0_PP",
      },
      Object {
        "available": true,
        "size": "P",
        "sku": "5807_343_0_P",
      },
      Object {
        "available": true,
        "size": "M",
        "sku": "5807_343_0_M",
      },
      Object {
        "available": true,
        "size": "G",
        "sku": "5807_343_0_G",
      },
      Object {
        "available": false,
        "size": "GG",
        "sku": "5807_343_0_GG",
      },
    ]
  }
  style="20002605"
>
  <styled.div
    onClick={[Function]}
    sale={false}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "dMvMzd",
            "rules": Array [
              "
  display: flex;

  &:hover {
    cursor: pointer;
  }

  .content {
    border: 1px solid ",
              "#eee",
              ";
    position: relative;
    flex: 1;

    &:hover {
      border-color: #ccc;
      transition: 0.2s all ease-in-out;
    }
  }

  .sale {
    position: absolute;
    top: 0;
    left: 0;
    background: ",
              "#000",
              ";
    color: ",
              "#fff",
              ";
    padding: 10px;
    text-align: center;

    .title {
      margin-bottom: 7px;
      text-decoration: underline;
    }
  }

  .image {
    position: relative;

    img {
      width: 100%;
      object-fit: cover;
    }
  }

  .bottomContent {
    padding: 5px 10px 15px;
  }

  .name {
    font-size: 20px;
    margin-bottom: 10px;
  }

  .prices {
    display: inline;
    margin-right: 10px;
  }

  .regularPrice {
    ",
              [Function],
              "
  }

  .actualPrice {
    color: ",
              ";
    font-weight: normal;

    ",
              [Function],
              "
  }

  .installments {
    color: #bbb;
    font-size: 12px;
    position: relative;
    top: -2px;
  }

  .size {
    margin-top: 15px;
    border: 1px solid #ddd;
    text-align: center;
    width: 25px;
    display: inline-block;
    margin-right: 5px;
    padding: 0px 0px 8px;

    span {
      font-size: 10px;
    }
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "usesTheme": false,
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      onClick={[Function]}
      sale={false}
    >
      <div
        className="sc-bdVaJa dMvMzd"
        onClick={[Function]}
      >
        <div
          className="content"
        >
          <div
            className="image"
          >
            <img
              alt="Foto do produto VESTIDO TRANSPASSE BOW"
              src="https://d3l7rqep7l31az.cloudfront.net/images/products/20002605_615_catalog_1.jpg?1460136912"
            />
          </div>
          <div
            className="bottomContent"
          >
            <p
              className="name"
            >
              VESTIDO TRANSPASSE BOW
            </p>
            <p
              className="regularPrice prices"
              key="regularPrice"
            >
              R$ 199,90
            </p>
            <p
              className="installments prices"
              key="installments"
            >
              3x R$ 66,63
            </p>
            <div
              className="sizes"
            >
              <div
                className="size"
                key="size-1"
              >
                <span>
                  P
                </span>
              </div>
              <div
                className="size"
                key="size-2"
              >
                <span>
                  M
                </span>
              </div>
              <div
                className="size"
                key="size-3"
              >
                <span>
                  G
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </StyledComponent>
  </styled.div>
</Product>
`;
